---
title: "Cellchat"
output: html_document
date: "2024-04-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# devtools::install_github("jinworks/CellChat")
# install.packages('NMF')
# devtools::install_github("jokergoo/circlize")
# devtools::install_github("jokergoo/ComplexHeatmap")
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("BiocNeighbors")
```

```{r}
library(CellChat)
library(patchwork)
library(Seurat)
library(NMF)
library(ggalluvial)
library(Seurat)
library(svglite)
options(stringsAsFactors = FALSE)
```

## Tutorial : [https://htmlpreview.github.io/?](https://htmlpreview.github.io/?https://github.com/jinworks/CellChat/blob/master/tutorial/CellChat-vignette.html#load-the-required-libraries){.uri}

<https://htmlpreview.github.io/?https://github.com/jinworks/CellChat/blob/master/tutorial/CellChat-vignette.html#load-the-required-libraries>

## [https://github.com/jinworks/CellChat/blob/master/tutorial/CellChat-vignette.html#load-the-required-libraries](https://htmlpreview.github.io/?https://github.com/jinworks/CellChat/blob/master/tutorial/CellChat-vignette.html#load-the-required-libraries){.uri}

```{r}
seurat.object <- readRDS("SCT/intgr_indiv_SCT_harmony/Subclustering/subcluster_annotated_RasV.rds")
```

```{r}
DimPlot(seurat.object, reduction = "umap", label = TRUE, group.by = "cell_subtype")
```

```{r}
cellchat <- createCellChat(object = seurat.object, group.by = "cell_subtype", assay = "SCT") ## needs the data layer (normilized values)

CellChatDB <- CellChatDB.mouse # use CellChatDB.human if running on mouse data
showDatabaseCategory(CellChatDB) # By default, the “Non-protein Signaling”isnt used
```

```{r}
dplyr::glimpse(CellChatDB$interaction)
cellchat@DB <- CellChatDB
```

```{r}
# subset the expression data of signaling genes for saving computation cost
cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database
future::plan("multisession", workers = 4) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)

#> The number of highly variable ligand-receptor pairs used for signaling inference is 692
#execution.time = Sys.time() - ptm
#print(as.numeric(execution.time, units = "secs"))

```

## Project gene expression data onto PPI (Optional: when running it, USER should set `raw.use = FALSE` in the function `computeCommunProb()` in order to use the projected data)

## For shallow sequencing depth use this

```{r}
# project gene expression data onto PPI (Optional: when running it, USER should set `raw.use = FALSE` in the function `computeCommunProb()` in order to use the projected data)
# for shallow sequencing depth  use this
cellchat <- projectData(cellchat, PPI.mouse)

```

```{r}
computeAveExpr(cellchat, type =  "truncatedMean", trim = 0.05)# this function calculates what cells are expressed in of signaling genes of interest so we can find a threshold

options(future.globals.maxSize= 891289600)
# calculate communication probabilities and infer cellular communication network
cellchat <- computeCommunProb(cellchat, raw.use = TRUE) # use False if you want to use protein projected data(for samples with shallow sequencing depth), also, by default , (type =  "truncatedMean", trim = 0.25) 25% so we can change that based on avg

```

```{r}
# filter out cc communication only if there are a few cells interacting
cellchat <- filterCommunication(cellchat, min.cells = 10) #10 uis default
```

```{r}
df.net <- subsetCommunication(cellchat)
#focus at the level of ligands/receptors. Set slot.name = "netP" to access the the inferred communications at the level of signaling pathways
df1.net <- subsetCommunication(cellchat, sources.use = c("Mono-like","Mph","Mo"), targets.use = c("CD8T","CD4T"))
```

## Infer CCC at a signaling pathway **level**

```{r}
# infer ccc pathway at a signaling pathway level
cellchat <- computeCommunProbPathway(cellchat)
cellchat <- aggregateNet(cellchat)
saveRDS(cellchat, file = "cellchat_RasV_1.rds")
```

```{r}
groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(2,2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
```

```{r}
#cellchat <- loadR
pathways.show.all <- cellchat@netP$pathways
pathways.show.all
```

```{r}
netAnalysis_signalingRole_scatter(cellchat, signaling = "Cholesterol")
```
## Take out the important subtypes we want to look
```{r}

cell_types <- unique(cellchat@meta$cell_subtype)

# Filter cell types
mph <- grep("_Mph$", cell_types, value = TRUE)
t <- grep("_T$", cell_types, value = TRUE)
b <- grep("_B$", cell_types, value = TRUE)
ec <- grep("_EC$", cell_types, value = TRUE)
kc <- grep("_KCs$", cell_types, value = TRUE)


par(mfrow=c(1,1))
netVisual_aggregate(cellchat, signaling = pathways.show, layout ="circle", color.use = NULL, sources.use = sources_use, targets.use = targets_use, idents.use = NULL)
```

```{r}
cellchat@net$count
cellchat@net$weight
#cellchat@netP$pathways # to access all pathways
```

```{r}
groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
```

## Part III: Visualization of cell-cell communication network

```{r}
mat <- cellchat@net$weight
par(mfrow = c(3,5), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
```

```{r}
cellchat <- readRDS(file = "cellchat_RasV_final.rds")
# Access all the signaling pathways showing significant communications
pathways.show.all <- cellchat@netP$pathways
pathways.show.all

# check the order of cell identity to set suitable vertex.receiver
levels(cellchat@idents)
vertex.receiver = seq(1,19)

for (i in 1:length(pathways.show.all)) {
  # Visualize communication network associated with both signaling pathway and individual L-R pairs
  netVisual(cellchat, signaling = pathways.show.all[i], vertex.receiver = vertex.receiver, layout = "hierarchy",out.format ="png",height =10)
  # Compute and visualize the contribution of each ligand-receptor pair to the overall signaling pathway
  gg <- netAnalysis_contribution(cellchat, signaling = pathways.show.all[i])
  ggsave(filename=paste0(pathways.show.all[i], "_L-R_contribution.pdf"), plot=gg, width = 3, height = 2, units = 'in', dpi = 300)
}



```

## **Visualize cell-cell communication mediated by multiple ligand-receptors or signaling pathways**

```{r}
# (1) show all the significant interactions (L-R pairs) from some cell groups (defined by 'sources.use') to other cell groups (defined by 'targets.use')

gg2 <- netVisual_bubble(cellchat, sources.use = 11, targets.use = c(0:11), remove.isolate = FALSE)
ggsave(filename= "Sugnificantinteractions.pdf", plot=gg2, width = 8, height = 10, units = 'in', dpi = 300)
```

```{r}

x <- c(2,5,3,9,8,11,6)
count <- 0
for (val in x) {
if(val %% 2 == 0)  count = count+1
}
print(count)
```
### Here to plot the LR interactions
```{r}

# (2) show all the significant interactions (L-R pairs) associated with certain signaling pathways
plot<- netVisual_bubble(cellchat, sources.use = t, targets.use = ec, signaling = pathways.show.all, remove.isolate = FALSE)
#> Comparing communications on a single object
ggsave(filename= "T_to_EC_LR.pdf", plot=plot, width = 10, height = 15, units = 'in', dpi = 300)

```

```{r}

pairLR.use <- extractEnrichedLR(cellchat, signaling = pathways.show.all)
netVisual_bubble(cellchat, sources.use = t, targets.use = ec, pairLR.use =pairLR.use, remove.isolate = TRUE)
```
## gene expression enrichment on pathways
```{r}
#
a <-plotGeneExpression(cellchat, signaling = pathways.show.all[1], enriched.only = TRUE, type = "violin")
a
ggsave(file="Colagen_percondition.pdf",a, width = 10, height =8 , dpi=300)
```

## Part IV: Systems analysis of cell-cell communication network

```{r}

# Compute the network centrality scores
cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
# Visualize the computed centrality scores using heatmap, allowing ready identification of major signaling roles of cell groups
for (i in 1:length(pathways.show.all)) {
    # Generate the plot
    svglite(paste0(pathways.show.all[i],'_heatmap_sa_cci.svg'),width = 10,heigh = 8)
    netAnalysis_signalingRole_network(cellchat, signaling = pathways.show.all[i], width = 19, height = 10, font.size = 8)
    dev.off()
}
```

```{r}
##Cholesterol, SPP1, ApoA/B/E,

# Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
gg1 <- netAnalysis_signalingRole_scatter(cellchat)
#> Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
# Signaling role analysis on the cell-cell communication networks of interest
gg2 <- netAnalysis_signalingRole_scatter(cellchat, signaling =c("Cholesterol","ApoA","ApoB","SPP1"))
#> Signaling role analysis on the cell-cell communication network from user's input

gg2
```

```{r}
# Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
ht1 <- netAnalysis_signalingRole_heatmap(cellchat, pattern = "outgoing")
ht2 <- netAnalysis_signalingRole_heatmap(cellchat, pattern = "incoming")
ht1
ggsave(file="Incoming_heatmap.pdf",ht1, width = 15, height =8 , dpi=300)
ggsave(file="Outgoing_heatmap.pdf",ht2, width = 22, height =8 , dpi=300)
```

```{r}
dplyr::glimpse(CellChatDB$interaction)
```



```{r}
# Seklect K shows you a curve and based on that select n patterns
selectK(cellchat, pattern = "incoming")
nPaterns <- 3
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "incoming", k= nPaterns, width = 5, height = 9)
inc<- netAnalysis_dot(cellchat, pattern= "incoming")
ggsave(file="incoming.pdf",inc, width = 22, height =8 , dpi=300)
```



```{r}
selectK(cellchat, pattern = "outgoing")
nPaterns <- 4
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing", k= nPaterns, width = 10, height = 8)
exc<-netAnalysis_dot(cellchat, pattern= "outgoing")
ggsave(file="outgoing.pdf",exc, width = 22, height =8 , dpi=300)
```

```{r}
netAnalysis_river(cellchat, pattern = "incoming")
netAnalysis_river(cellchat, pattern = "outgoing")
```

```{r}
cellchat <- computeNetSimilarity(cellchat, type = "functional")
cellchat <- netEmbedding(cellchat, type = "functional")
#> Manifold learning of the signaling networks for a single dataset
cellchat <- netClustering(cellchat, type = "functional")
#> Classification learning of the signaling networks for a single dataset
# Visualization in 2D-space
netVisual_embedding(cellchat, type = "functional", label.size = 3.5)
```

```{r}
cellchat <- computeNetSimilarity(cellchat, type = "structural")
cellchat <- netEmbedding(cellchat, type = "structural")
#> Manifold learning of the signaling networks for a single dataset
cellchat <- netClustering(cellchat, type = "structural")
#> Classification learning of the signaling networks for a single dataset
# Visualization in 2D-space
netVisual_embedding(cellchat, type = "structural", label.size = 3.5)
```

```{r}
saveRDS(cellchat, file = "cellchat_RasV_final.rds")
```

```{r}
#Part VI: Explore the cell-cell communication through the Interactive CellChat Explorer
runCellChatApp(cellchat)
```
